contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,594 @@
+
+        require(
+            amount >= _config.minimumContribution,
+            "StakingPool: min contribution"
+        );
+        require(
+            _totalStakedAmount + amount < _config.maxTotalPoolStake,
+            "StakingPool: oversubscribed"
+        );
+        //slither-disable-next-line timestamp
+        require(
+            block.timestamp < _config.epochStartTimestamp,
+            "StakingPool: too late"
+        );
+
+        User storage user = _users[_msgSender()];
+
+        user.depositAmount += uint128(amount);
+        _totalStakedAmount += uint128(amount);
+
+        emit Deposit(_msgSender(), amount);
+
+        // calculate/update rewards
+        if (_config.rewardType == StakingPoolLib.RewardType.FLOATING) {
+            _updateRewardsRatios(_config);
+        }
+        if (_config.rewardType == StakingPoolLib.RewardType.FIXED) {
+            _calculateFixedRewards(_config, user, amount);
+        }
+
+        require(
+            _config.stakeToken.transferFrom(
+                _msgSender(),
+                address(this),
+                amount
+            ),
+            "StakingPool: deposit tx fail"
+        );
+    }
+
+    /**
+     * @notice Withdraw both stake and reward tokens when the stake period is complete
+     */
+    function withdraw()
+        external
+        whenNotPaused
+        stakingPeriodComplete
+        rewardsAvailable
+        nonReentrant
+    {
+        User memory user = _users[_msgSender()];
+        require(user.depositAmount > 0, "StakingPool: not eligible");
+
+        delete _users[_msgSender()];
+
+        StakingPoolLib.Config storage _config = _stakingPoolConfig;
+
+        //slither-disable-next-line reentrancy-events
+        _transferStake(user.depositAmount, IERC20(_config.stakeToken));
+
+        _withdrawRewards(_config, user);
+    }
+
+    /**
+     * @notice Withdraw only stake tokens after staking period is complete. Reward tokens may not be available yet.
+     */
+    function withdrawStake()
+        external
+        stakingPeriodComplete
+        nonReentrant
+        whenNotPaused
+    {
+        User storage user = _users[_msgSender()];
+        require(user.depositAmount > 0, "StakingPool: not eligible");
+
+        uint128 currentDepositBalance = user.depositAmount;
+        user.depositAmount = 0;
+
+        StakingPoolLib.Config storage _config = _stakingPoolConfig;
+
+        // set users floating reward
+        for (uint256 i = 0; i < _config.rewardTokens.length; i++) {
+            if (_config.rewardType == StakingPoolLib.RewardType.FLOATING) {
+                user.rewardAmounts[i] = _calculateFloatingReward(
+                    _config.rewardTokens[i].ratio,
+                    currentDepositBalance
+                );
+            }
+        }
+
+        _transferStake(currentDepositBalance, IERC20(_config.stakeToken));
+    }
+
+    /**
+     * @notice Withdraw only reward tokens. Stake may have already been withdrawn.
```

Comment:
Previously there was a bug where if the user withdrew their rewards **before** their stake, their stake balance would be deleted. To prevent this we enforce that either rewards and stake are withdrawn at the same time, or the stake first, and then the rewards.

TLDR will change this comment to`Withdraw only reward tokens. Stake **must** have already been withdrawn.`

Reply:
Done in commit d1f1bf7